File: public/cpp/084-kth-largest-element.cpp
/*
 * Problem: Kth Largest Element in an Array (LeetCode 215)
 * Link: https://leetcode.com/problems/kth-largest-element-in-an-array/
 * Difficulty: Medium
 * Category: Heaps and Priority Queues
 *
 * Description:
 * Given an integer array nums and an integer k, return the kth largest element
 * in the array. Note that it is the kth largest element in the sorted order,
 * not the kth distinct element.
 * Can you solve it without sorting?
 *
 * Example 1: nums=[3,2,1,5,6,4], k=2  → 5
 * Example 2: nums=[3,2,3,1,2,4,5,5,6], k=4 → 4
 *
 * Constraints:
 * - 1 <= k <= nums.length <= 10^5
 * - -10^4 <= nums[i] <= 10^4
 */

#include <vector>
#include <algorithm>
#include <iostream>
#include <queue>
#include <random>
using namespace std;

/*
 * APPROACH 1: MIN HEAP OF SIZE K (OPTIMAL FOR SMALL K)
 *
 * Maintain a min-heap of size k.  The top of the heap is always the kth largest.
 * For each element: if heap < k → push; else if element > top → replace top.
 *
 * Time:  O(n log k)
 * Space: O(k)
 */
class Solution1 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> minHeap;
        for (int num : nums) {
            minHeap.push(num);
            if ((int)minHeap.size() > k) minHeap.pop();
        }
        return minHeap.top();
    }
};

/*
 * APPROACH 2: SORT DESCENDING
 *
 * Sort the array in descending order and return element at index k-1.
 *
 * Time:  O(n log n)
 * Space: O(1)  (in-place sort)
 */
class Solution2 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end(), greater<int>());
        return nums[k - 1];
    }
};

/*
 * APPROACH 3: QUICKSELECT (AVERAGE O(n))
 *
 * Partition the array around a pivot (Dutch National Flag style).
 * If pivot lands at position n-k we are done; otherwise recurse on
 * the correct side.  Uses random pivot to avoid worst-case O(n²).
 *
 * Time:  O(n) average, O(n²) worst
 * Space: O(log n) recursion stack
 */
class Solution3 {
    int partition(vector<int>& nums, int lo, int hi) {
        // Random pivot to avoid sorted-input worst case
        int pivotIdx = lo + rand() % (hi - lo + 1);
        swap(nums[pivotIdx], nums[hi]);
        int pivot = nums[hi], i = lo;
        for (int j = lo; j < hi; j++)
            if (nums[j] <= pivot) swap(nums[i++], nums[j]);
        swap(nums[i], nums[hi]);
        return i;
    }
    int quickselect(vector<int>& nums, int lo, int hi, int target) {
        if (lo == hi) return nums[lo];
        int p = partition(nums, lo, hi);
        if (p == target) return nums[p];
        return p < target ? quickselect(nums, p + 1, hi, target)
                          : quickselect(nums, lo, p - 1, target);
    }
public:
    int findKthLargest(vector<int>& nums, int k) {
        // kth largest = element at index n-k in sorted ascending order
        return quickselect(nums, 0, nums.size() - 1, nums.size() - k);
    }
};

/*
 * APPROACH 4: MAX HEAP (SIMPLE, EXTRACT K TIMES)
 *
 * Build a max-heap, pop k times.
 *
 * Time:  O(n + k log n)
 * Space: O(n)
 */
class Solution4 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int> maxHeap(nums.begin(), nums.end());
        for (int i = 0; i < k - 1; i++) maxHeap.pop();
        return maxHeap.top();
    }
};

/*
 * APPROACH 5: COUNTING SORT  (when values bounded)
 *
 * Because constraints guarantee -10^4 <= nums[i] <= 10^4, we can bucket-count.
 *
 * Time:  O(n + R)  where R = value range = 20001
 * Space: O(R)
 */
class Solution5 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        const int OFFSET = 10000;
        const int R = 20001;
        vector<int> cnt(R, 0);
        for (int n : nums) cnt[n + OFFSET]++;
        int remain = k;
        for (int v = R - 1; v >= 0; v--) {
            remain -= cnt[v];
            if (remain <= 0) return v - OFFSET;
        }
        return -1; // unreachable
    }
};

void test(vector<int> nums, int k, int approach) {
    int result;
    cout << "[";
    for (int i = 0; i < (int)nums.size(); i++) { cout << nums[i]; if (i+1<(int)nums.size()) cout<<","; }
    cout << "] k=" << k << "  ";
    switch(approach) {
        case 1: { Solution1 s; result=s.findKthLargest(nums,k); cout<<"MinHeap:     "; break; }
        case 2: { Solution2 s; result=s.findKthLargest(nums,k); cout<<"Sort:        "; break; }
        case 3: { Solution3 s; result=s.findKthLargest(nums,k); cout<<"Quickselect: "; break; }
        case 4: { Solution4 s; result=s.findKthLargest(nums,k); cout<<"MaxHeap:     "; break; }
        case 5: { Solution5 s; result=s.findKthLargest(nums,k); cout<<"CountSort:   "; break; }
    }
    cout << result << "\n";
}

int main() {
    for(int i=1;i<=5;i++) test({3,2,1,5,6,4},2,i);
    cout << "\n";
    for(int i=1;i<=5;i++) test({3,2,3,1,2,4,5,5,6},4,i);
    cout << "\n";
    for(int i=1;i<=5;i++) test({1},1,i);
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (Min Heap k):   Time O(n log k)        Space O(k)
 * Approach 2 (Sort):         Time O(n log n)         Space O(1)
 * Approach 3 (Quickselect):  Time O(n) avg           Space O(log n)
 * Approach 4 (Max Heap):     Time O(n + k log n)     Space O(n)
 * Approach 5 (Count Sort):   Time O(n + R)           Space O(R)
 */

// Made with Bob

File: src/data/tutorials/084-kth-largest-element.js
/**
 * Tutorial content for #084 — Kth Largest Element in an Array
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an integer array nums and an integer k, return the kth largest element in the array. Note: it is the kth largest in sorted order, not the kth distinct element. For example, k=1 means the maximum, k=2 means the second largest, and so on.`,
    example: `nums = [3,2,1,5,6,4], k = 2\nSorted descending: [6,5,4,3,2,1]\n2nd element = 5\n✅ Answer: 5`,
    keyInsight: `You don't need the full sorted array — just the k-th position. A min-heap of size k does this in O(n log k): keep the k largest elements seen so far, discarding smaller ones. Or use Quickselect for O(n) average.`,
  },

  approaches: {
    'Min Heap of Size K': {
      intuition: `Keep a min-heap that holds exactly k elements. As you scan the array, add each element to the heap. Whenever the heap grows beyond k, remove the smallest element (the heap's minimum). After the full scan, the heap holds the k largest elements, and the minimum of the heap IS the kth largest.`,
      steps: [
        `Create an empty min-heap.`,
        `For each num in nums:`,
        `  Add num to the heap.`,
        `  If heap.size > k: remove the smallest (heap.poll()).`,
        `Return heap.peek() — the smallest among the k largest.`,
      ],
      example: `nums=[3,2,1,5,6,4], k=2\n\nadd 3: heap:[3]\nadd 2: heap:[2,3]\nadd 1: size=3>2 → remove min(1). heap:[2,3]\nadd 5: heap:[2,3,5] → remove 2. heap:[3,5]\nadd 6: heap:[3,5,6] → remove 3. heap:[5,6]\nadd 4: heap:[4,5,6] → remove 4. heap:[5,6]\nReturn heap.peek() = 5 ✅`,
      keyInsight: `O(n log k) time, O(k) space. For small k this is faster than sorting the entire array. This is the go-to interview solution.`,
    },

    'Sort Descending': {
      intuition: `The simplest approach: sort the entire array in descending order and return the element at index k-1. No heap, no selection — just sort and index.`,
      steps: [
        `Sort nums in descending order.`,
        `Return nums[k - 1].`,
      ],
      example: `nums=[3,2,1,5,6,4], k=2\nSorted desc: [6,5,4,3,2,1]\nnums[k-1] = nums[1] = 5 ✅`,
      keyInsight: `O(n log n) time, O(1) extra space (in-place sort). Simple to write but wasteful — you sort the entire array to find just one element. Acceptable for small inputs or when code clarity matters most.`,
    },

    'Quickselect': {
      intuition: `Quickselect is a partial sort: it finds the kth smallest/largest in O(n) average time without fully sorting the array. Pick a pivot, partition the array so everything larger than the pivot is to the left (for kth largest). If the pivot lands at position k-1, you're done. Otherwise recurse on the appropriate half.`,
      steps: [
        `Convert "kth largest" to "find the element at index k-1 after descending sort", or equivalently index n-k after ascending sort.`,
        `Partition: pick a pivot, place all elements > pivot to the left, < pivot to the right.`,
        `If pivot index == target index: return pivot.`,
        `If pivot index < target: recurse on the right sub-array.`,
        `Else: recurse on the left sub-array.`,
      ],
      example: `nums=[3,2,1,5,6,4], k=2. Target index=1 (0-based).\n\nPartition around pivot=4: [6,5,4,3,2,1] (desc order)\nPivot 4 lands at index 2. Target=1, so go left subarray.\nPartition [6,5] around 5: [6,5]. Pivot lands at index 1. Match!\nReturn 5 ✅`,
      keyInsight: `O(n) average, O(n²) worst-case time, O(1) extra space. Fastest on average but the worst case can be mitigated with random pivot selection. Great for large arrays where O(n log n) sort is too slow.`,
    },

    'Max Heap (Extract k Times)': {
      intuition: `Build a max-heap from all elements, then extract the maximum k times. The kth extraction gives you the kth largest element.`,
      steps: [
        `Heapify all nums into a max-heap.`,
        `Call heap.extractMax() k times.`,
        `Return the last extracted value.`,
      ],
      example: `nums=[3,2,1,5,6,4], k=2\n\nMax-heap: [6,5,4,3,2,1]\nExtract 1: 6. heap:[5,4,3,2,1]\nExtract 2: 5. ← 2nd largest\nReturn 5 ✅`,
      keyInsight: `O(n + k log n) time, O(n) space. Building the heap is O(n), each extraction is O(log n). Worse than the Min Heap of size k approach for large n and small k.`,
    },

    'Counting Sort': {
      intuition: `If the range of values is bounded (e.g., -10000 to 10000 per LeetCode constraints), use a frequency array. Count how many times each value appears. Then scan from the maximum value downward, decrementing a counter by each frequency, until the counter drops below k.`,
      steps: [
        `Find min and max of nums to determine range.`,
        `Build a frequency array count[].`,
        `For each value in nums: count[val - offset]++.`,
        `Scan from the highest value downward: k -= count[val]. When k ≤ 0, return val.`,
      ],
      example: `nums=[3,2,1,5,6,4], k=2. Range 1–6.\ncount: [1,1,1,1,1,1] (each appears once)\n\nval=6: k=2-1=1. not ≤0.\nval=5: k=1-1=0. k≤0 → return 5 ✅`,
      keyInsight: `O(n + range) time, O(range) space. Extremely fast when the value range is small and fixed. Not useful for large or unbounded value ranges.`,
    },
  },
}
